File: leetcode2/2medium/노형민/Q808.cs
public class Solution
{
    double?[,] possibility;

    public double SoupServings(int n)
    {
        n = (n+24) / 25;
        
        if (n > 4800)
        {
            return 1.0;
        }

        possibility = new double?[n + 1, n + 1];

        return GetPossibility(n, n);
    }

    private double GetPossibility(int i, int j)
    {
        if (i <= 0 && j <= 0)
        {
            return 0.5;
        }
        if (i <= 0)
        {
            return 1.0;
        }
        if (j <= 0)
        {
            return 0.0;
        }

        if (possibility[i, j] != null)
        { 
            return possibility[i, j].Value;
        }

        double result = 0.25 * (
            GetPossibility(i - 4, j) +
            GetPossibility(i - 3, j - 1) +
            GetPossibility(i - 2, j - 2) +
            GetPossibility(i - 1, j - 3)
        );

        possibility[i, j] = result;

        return result;
    }
}

File: leetcode2/2medium/최원준/Q808.java
package Leetcode.최원준;

/*
1. 아이디어 :


2. 시간복잡도 :
O( )

3. 자료구조/알고리즘 :

 */

public class Q808 {
    class Solution {
        Double[][] cache;
        private double dfs(int A, int B) {
            if (A<=0 && B>0) return 1;
            if (A<=0 && B<=0) return 0.5;
            if (A>=0 && B<=0) return 0;
            if (cache[A][B] != null) return cache[A][B];

            double total = 0;
            total += dfs(A-4, B-0);
            total += dfs(A-3, B-1);
            total += dfs(A-2, B-2);
            total += dfs(A-1, B-3);
            cache[A][B] = total/4;
            return total/4;
        }
        public double soupServings(int n) {
            if (n>=4500) return 1;
            cache = new Double[n/25+2][n/25+2];
            return dfs((int)Math.ceil(n/25.0), (int)Math.ceil(n/25.0));
        }
    }
}
